format-delimited/src/main/java/io/cdap/plugin/format/delimited/common/DataTypeDetectorUtils.java

```diff
@@ -141,11 +142,57 @@ public static void validateSchemaFieldNames(String[] fieldNames) {
         throw new IllegalArgumentException(
           String.format(
             "Invalid column name detected: \"%s\"! " +
-            "Column names can only start with capital letters, small letters, and " + "\"_\". " +
-            "Subsequently they may contain only capital letters, small letters, numbers  and \"_\".", fieldName
+              "Column names can only start with capital letters, small letters, and " + "\"_\". " +
+              "Subsequently they may contain only capital letters, small letters, numbers  and \"_\".", fieldName
           )
         );
       }
     }
   }
+
+  /**
+   * Cleans an array of field names to make sure they comply with avro field naming standard.
+   * It also makes sure each name is unique in the list.
+   * Field names can start with [A-Za-z_] and subsequently contain only [A-Za-z0-9_].
+   *
+   * Steps:
+   * 1) Trim surrounding spaces
+   * 2) If it starts with a number, prepend "_"
+   * 3) Replace invalid characters with "_"
+   * 4) Check if the name has been found before if so add _# where # is the number of times seen before
+   * @param fieldNames an array of field names to be cleaned
+   */
+  private static String[] cleanSchemaFieldNames(String[] fieldNames) {
+    final String replacementChar = "_";
+    final List<String> cleanFieldNames = new ArrayList<>();
+    final Map<String, Integer> seenFieldNames = new HashMap<>();
+    final Pattern notValidPattern = Pattern.compile("[^A-Za-z0-9_]");
```

Comment:
Please add a constant with precompiled pattern. Consider adding "+" to the pattern as usually it makes sense to replace all invalid chars with a single _. Otherwise you end up with _, __, ___. I'd prefer _1,_2,_3 much better :)